infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/CommonPVCStrategy.java

```diff
@@ -258,4 +274,70 @@ private PersistentVolumeClaim replacePVCsWithCommon(
         .filter(subpath -> !isNullOrEmpty(subpath))
         .collect(Collectors.toSet());
   }
+
+  private void deleteCommonPVC(Workspace workspace) throws InfrastructureException {
+    factory.get(workspace).persistentVolumeClaims().delete(configuredPVCName);
+  }
+
+  /**
+   * @return true, if the common PVC is expected to be deleted, false otherwise. Depending on the
+   *     configuration it could be either the common PVC of a particular user, or the common PVC
+   *     that is shared across all the users (the last setup is considered to be uncommon and
+   *     not-recommended)
+   */
+  private boolean noWorkspacesLeft() {
+    return defaultNamespaceWithoutPlaceholderIsDefined()
+        ? totalNumberOfWorkspacesIsZero()
+        : userHasNoWorkspaces();
+  }
+
+  /**
+   * The method is expected to be used in order to identify if the common PVC is used across all the
+   * users. The common PVC for all the users will be used if
+   * 'che.infra.kubernetes.namespace.default' points to a particular namespace e.g. 'che',
+   * 'workspaces' etc.
+   *
+   * @return true, if 'che.infra.kubernetes.namespace.default' is defined and does NOT contain the
+   *     placeholder e.g.: che-workspace-<username>), false otherwise
+   */
+  private boolean defaultNamespaceWithoutPlaceholderIsDefined() {
+    return defaultNamespaceName != null
+        && !defaultNamespaceName.contains("<")
+        && !defaultNamespaceName.contains(">");
+  }
+
+  /** @return true, if there are no workspaces left across all the users, false otherwise */
+  private boolean totalNumberOfWorkspacesIsZero() {
+    try {
+      return workspaceManager.getWorkspacesTotalCount() == 0;
+    } catch (ServerException e) {
+      log.error("Unable to get the total number of workspaces. Cause: {}", e.getMessage(), e);
+    }
+    return false;
+  }
+
+  /**
+   * @return true, if a given user has no workspaces, false otherwise (or if the subject is
+   *     anonymous)
+   */
+  private boolean userHasNoWorkspaces() {
+    Subject subject = EnvironmentContext.getCurrent().getSubject();
+    if (!subject.isAnonymous()) {
+      String userId = subject.getUserId();
+      try {
+        Page<WorkspaceImpl> workspaces = workspaceManager.getWorkspaces(userId, false, 1, 0);
+        if (workspaces.isEmpty()) {
+          log.debug("User '{}' has no more workspaces left", userId);
+          return true;
+        }
+      } catch (ServerException e) {
+        log.error(
```

Comment:
This is an impossible sutuation. I believe we should throw a RuntimeException here. Or remove this block at all.

Reply:
I now throw a RuntimeException here